test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala

```diff
@@ -67,174 +32,380 @@ class FormatterTest {
+    assertF("D431", "%X", 54321)
+    assertF("FFFF2BCF", "%X", -54321)
+
+    assertF("0xd431", "%#x", 54321)
+    assertF("0xffff2bcf", "%#x", -54321)
+    assertF("0XD431", "%#X", 54321)
+    assertF("0XFFFF2BCF", "%#X", -54321)
+
+    assertF("         d431", "%13x", 54321)
+    assertF("     ffff2bcf", "%13x", -54321)
+    assertF("         D431", "%13X", 54321)
+    assertF("     FFFF2BCF", "%13X", -54321)
+    assertF("       0xd431", "%#13x", 54321)
+    assertF("   0xffff2bcf", "%#13x", -54321)
+
+    assertF("d431         ", "%-13x", 54321)
+    assertF("ffff2bcf     ", "%-13x", -54321)
+    assertF("0xd431       ", "%-#13x", 54321)
+    assertF("0xffff2bcf   ", "%-#13x", -54321)
+
+    assertF("000000000d431", "%013x", 54321)
+    assertF("00000ffff2bcf", "%013x", -54321)
+    assertF("0x0000000d431", "%0#13x", 54321)
+    assertF("0x000ffff2bcf", "%0#13x", -54321)
+
+    assertF("fffffffc", "%x", asIntOnJVM(-4.toByte))
+    assertF("0x005", "%0#5x", 5.toByte)
+    assertF("  0x5", "%#5x", 5.toByte)
+    assertF("  0X5", "%#5X", 5.toByte)
+    assertF("fffffffd", "%x", asIntOnJVM(-3.toByte))
+
+    assertF("0x005", "%0#5x", 5.toShort)
+    assertF("  0x5", "%#5x", 5.toShort)
+    assertF("  0X5", "%#5X", 5.toShort)
+    assertF("fffffffd", "%x", asIntOnJVM(-3.toShort))
+
+    assertF("ffffffffffff2bcf", "%x", -54321L)
+    assertF("28EEA4CB1", "%X", 10987654321L)
   }
 
   @Test def `should_provide_e_conversion`(): Unit = {
-    assertF("1.000000e+03", "%e", new JDouble(1000))
-    assertF("1e+100", "%.0e", new JDouble(1.2e100))
-    // We use 1.51e100 in this test, since we seem to have a floating
-    // imprecision at exactly 1.5e100 that yields to a rounding error
-    // towards (1e+100 instead of 2e+100)
-    assertF("2e+100", "%.0e", new JDouble(1.51e100))
-    assertF(" 1.20e+100", "%10.2e", new JDouble(1.2e100))
-    assertF("001.2000e-21", "%012.4e", new JFloat(1.2e-21f))
-    assertF("001.2000E-21", "%012.4E", new JFloat(1.2e-21f))
-    assertF("(0001.2000e-21)", "%(015.4e", new JFloat(-1.2e-21f))
-
-    // Tests with infinity and NaN
-    assertF("Infinity", "%e", new JDouble(Double.PositiveInfinity))
-    assertF("-Infinity", "%e", new JDouble(Double.NegativeInfinity))
-    assertF("  Infinity", "%010e", new JDouble(Double.PositiveInfinity))
-    assertF("Infinity  ", "%-10e", new JDouble(Double.PositiveInfinity))
-    assertF("(Infinity)", "%(e", new JDouble(Double.NegativeInfinity))
-    assertF("       NaN", "%010e", new JDouble(Double.NaN))
+    assertF("1.000000e+03", "%e", 1000.0)
+    assertF("1e+100", "%.0e", 1.2e100)
+    assertF("0.000e+00", "%.3e", 0.0)
+
+    /* We use 1.51e100 in this test, since we seem to have a floating point
+     * imprecision at exactly 1.5e100 that yields to a rounding error towards
+     * 1e+100 instead of 2e+100.
+     */
+    assertF("2e+100", "%.0e", 1.51e100)
+
+    assertF(" 1.20e+100", "%10.2e", 1.2e100)
+
+    // Corner case of round-half-up (currently non-compliant)
```

Comment:
Should we track this as a bug?

Reply:
I explicitly generalized #3202 for this set of issues.